### ISSUE TITLE
Support JPA Entity Lifecycle Callbacks

### ISSUE THREAD

**Issue by AlbozDroid** [quarkusio/quarkus#12740]
**Description**
Implement JPA Callbacks

JPA defines the following Entity lifecycle callbacks:
`@PrePersist` - before a new entity is persisted (added to the EntityManager).
`@PostPersist` - after storing a new entity in the database (during commit or flush).
`@PostLoad` - after an entity has been retrieved from the database.
`@PreUpdate` - when an entity is identified as modified by the EntityManager.
`@PostUpdate` - after updating an entity in the database (during commit or flush).
`@PreRemove` - when an entity is marked for removal in the EntityManager.
`@PostRemove` - after deleting an entity from the database (during commit or flush).

Hibernate implements them all, however, they are not supported in Quarkus according to:

 https://quarkus.io/guides/hibernate-orm#limitations.

They are useful to schedule events and actions. As an example, we use the `@PostPersist` and `@PostUpdate` callbacks to keep in sync the database and elasticsearch. The DB is the single source of truth and when something changes in the DB the record should be refresh in the search engine (Elastic).

The above is just an example, but there are many others too.

**Implementation ideas**

The implementations already exist in Hibernate.

**Comment by famod** [comment]
Hm, this documentation bit seems outdated. Both `@PrePersist` and `@PreUpdate` are working as expected in my/our Quarkus app.

/cc @Sanne @gsmet

**Comment by Sanne** [comment]
@famod I'm pretty sure they won't work in native, and users will also be likely to expect CDI integrations to work.

**Comment by famod** [comment]
@Sanne I see. Maybe we should refine the docs?
I and my team are not the only ones using those callbacks already (see other issues).

**Comment by Sanne** [comment]
Sure we could refine the docs, but it bothers me to have some funtionality which only works in JVM mode; I'd much rather keep it simple and document this is not supported yet - if it works for you you're "trespassing" the intent :)

It would set a bad precedent for other extension developers to have some features that behave differently in native-images - everything becomes more complex if each sub-feature needs a flag "works in native", we strived from the beginning to send a clear message: Quarkus apps which work in JVM will work the same in native.

This principle becomes important when you find yourself in need of a native debugger: most Java developers would rather not need to see one ever, so it's rather important that an application is semantically consistent between its JVM mode and its native mode - it really helps to keep the differences to a minimal, so that most issues can be debugged in JVM mode, then compiled to native later.

Maybe to be coherent with this principle I should make sure people don't use callbacks in JVM mode either :) But I'm trying to be pragmatic here, and I guess we should rather spend our time to fix this limitation in native. To be fair I wanted this done months ago, I have to apologies we've been a bit overwhelmed by our own success - I'll see what we can do.

**Comment by AlbozDroid** [comment]
Personally, I think Quarkus should strive to have full support for **Microprofile** specs and **Hibernate** (fully JPA compliant). They're the backbone of Java Web Apps and the only serious alternative to Spring/SpringBoot.

**Comment by Sanne** [comment]
@AlbozDroid agreed, and no worries we'll implement these JPA callbacks. It's a limitation we inherited from the very initial POC, we simply never had the time but it's not inherently too difficult :)

To me it was also interesting that nobody had opened this issue yet ;-) It's a strategy to prioritieze things.

**Comment by famod** [comment]
@Sanne 

I totally get the "it shall work in native" goal.

> Maybe to be coherent with this principle I should make sure people don't use callbacks in JVM mode either

Please don't! If you do I am sure you will see more issues than you want. 😉

> To me it was also interesting that nobody had opened this issue yet

There is another one, more or less: #6948

> I have to apologize we've been a bit overwhelmed by our own success

Not the worst thing to happen I would say.

**Comment by Sanne** [comment]
thanks @famod ! Sure no worries I'm not disabling it yet, we are otherwise busy :)

**Comment by stuartwdouglas** [comment]
What is the issue with native here? Just a matter of registering them for reflection?

**Comment by Sanne** [comment]
@stuartwdouglas probably yes, I haven't looked at it in a while. AFAIR the whole handling of JPA callbacks needs a general refactoring: it entirely relies on runtime annotation processing, reflective bean instantiation, etc..  and there's probably some ArC integration to do as well.

"Just" adding a list of stuff to reflection could do the trick, but it also implies we need to keep a significant chunk of metadata held in memory so I would hope someone could dedicate some days to explore the better solution.

**Comment by kucharzyk** [comment]
Does Quarkus we have any alternatives to invoke some extra logic after persisting entity?

My use case is saving basic informations about time and user who create and modify entity. Using enverse here would be overkill. I don't want also do it manually. 

For dates I can use Hibernate @CreationTimestamp and @UpdateTimestamp. That works with Quarkus in native mode.

But what can I do with username of user who created or updated entity. Does Quarkus provide easy way to add interceptors without manually creating SessionFactory?

**Comment by jglink** [comment]
Are there any plans on implementing the JPA callback support for native mode in the near future?

**Comment by gavinking** [comment]
Would the old legacy Hibernate `Interceptor` work here?

**Comment by kucharzyk** [comment]
@gavinking @Sanne 
I've checked if Hibernate Interceptors works in native mode and my results are really suprising .
With Quarkus 1.13.1.Final both ```Interceptor``` and ```@PrePersist` WORKS IN NATIVE MODE.

Maybe this issue was already fixed ( intentionally or not)

**Comment by jglink** [comment]
I played a bit more with my simple test case.
If you annotate the JPA callback class with `@RegisterForReflection` it works quite nicely.

Could someone from the Quarkus team comment on this if this is an expected behavior and if this approach is safe to use.
Thanks

**Comment by Sanne** [comment]
yes the pending issue relates with the missing reflective rules.

So I believe `@RegisterForReflection` is a reasonable workaround; I plan to make it unnecessary in the future though.

**Comment by Sanne** [comment]
BTW the solution I'm aiming for does not automatically add the reflection rules; I'm aiming for a more optimal solution.

**Comment by kucharzyk** [comment]
@jglink I've cleaned project and tried again in native mode. It works.

Maybe you have something ususual in your setup.

My classes:

```java
@Entity
@Table(name = "users")
public class User {

    protected User() {
    }

    public User(String name) {
        this.name = name;
    }

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    private String name;
    
    private String createdDate;
    
    @PrePersist
    public void prePersist(){
        this.createdDate = "today";
    }

    // getters, setters, toString()
}
```

```java
@Path("/test")
@Transactional
public class ReactiveGreetingResource {

    @Inject
    EntityManager entityManager;

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Blocking
    public User test() {

        User user = new User(UUID.randomUUID().toString());
        entityManager.persist(user);
        return user;
    }
}
```

```
# datasource configuration
quarkus.datasource.db-kind = postgresql
quarkus.datasource.username = test
quarkus.datasource.password = test
quarkus.datasource.jdbc.url = jdbc:postgresql://localhost:5432/test

# drop and create the database at startup (use `update` to only update the schema)
quarkus.hibernate-orm.database.generation=drop-and-create
```

Really simple example but It works without any issues

**Comment by jglink** [comment]
@kucharzyk 
In your case this works out-of-the-box because you have the @PrePersist added to a method in the entity itself.
In my case, I had an own class with the @PrePersist to use it for all of my entities. And for this I had to annotate it with the @RegisterForReflection annotation.

**Comment by kucharzyk** [comment]
You have right @jglink in '''@PrePersit''' now works in entity but '''@EntityListeners''' needs '''@RegisterForReflection''' annotation.

**Comment by geoand** [comment]
@Sanne as there seems to be plenty of interest in this, maybe initially we should go for the register-for-reflection solution and then at some later time opt for an optimal solution?

**Comment by Sanne** [comment]
@geoand people can already do this? Just register all needed components for reflection.

We should make it work properly though.

**Comment by geoand** [comment]
Well my point is that they should not have to do it. Same as they don't have to for example when returning POJOs from REST endpoints or message processors.

I completely understand that a proper solution would be ideal, but if something is pretty highly requested and easily fixable, should we not provide that fix until time permits us to come up with a better solution?

**Comment by Sanne** [comment]
Sure, it's not far down on our TODO list.
